src/importer.cpp

```diff
@@ -150,130 +125,258 @@ bool Importer::ImporterImpl::doResolveImports(ModelPtr &model, const std::string
+            return false;
+        }
+        std::stringstream buffer;
+        buffer << file.rdbuf();
+        auto parser = Parser::create();
+        model = parser->parseModel(buffer.str());
+        mLibrary.insert(std::make_pair(url, model));
+    } else {
+        model = mLibrary[url];
+    }
+    importSource->setModel(model);
+    return true;
+}
+
+bool Importer::ImporterImpl::fetchImportSource(const ImportSourcePtr &importSource, Type type, const std::string &baseFile, std::vector<std::tuple<std::string, std::string, std::string>> &history)
+{
+    // If the model has never been retrieved, get it and add to library.
+    if (!importSource->hasModel()) {
+        if (!fetchModel(importSource, baseFile)) {
+            return false;
+        }
+    }
+    // Check for cycles.
+    std::string url = importSource->url();
+    if (mLibrary.count(url) == 0) {
+        url = resolvePath(url, baseFile);
+    }
+    if (!checkForCycles(mLibrary[url], history)) {
+        makeIssueCyclicDependency(mLibrary[url], type, history);
+        return false;
+    }
+    return true;
+}
+
+bool Importer::ImporterImpl::fetchComponent(const ComponentPtr &importComponent, const std::string &baseFile, std::vector<std::tuple<std::string, std::string, std::string>> &history)
+{
+    // Given the importComponent, check whether it has been resolved previously.  If so, return.
+    // If not, check for model, and parse/instantiate/add to library if needed.
+    // If model exists, resolve component's requirements, including child components and units required.
+
+    if (!importComponent->requiresImports() || importComponent->isResolved()) {
+        return true;
+    }
+
+    if (!importComponent->isImport()) {
+        // This component is not an import, but a descendant is.
+        for (size_t c = 0; c < importComponent->componentCount(); ++c) {
+            if (!fetchComponent(importComponent->component(c), baseFile, history)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    history.emplace_back(std::make_tuple(importComponent->name(), importComponent->importReference(), importComponent->importSource()->url()));
+    if (!fetchImportSource(importComponent->importSource(), Type::COMPONENT, baseFile, history)) {
+        return false;
+    }
+
+    // Check that the model instance in the library has resolved all of the required dependencies.
+    auto sourceModel = importComponent->importSource()->model();
+    auto sourceComponent = sourceModel->component(importComponent->importReference());
+    if (sourceComponent != nullptr) {
+        // Check whether the sourceComponent is itself an import. Note that the file path passed in
+        // here must be the path to the sourceModel, rather than the path to the previous import, so that
+        // the chain always tests local to the importing model first.
+        auto newBase = baseFile + directoryPath(importComponent->importSource()->url());
```

Comment:
shouldn't this be using `resolvePath` ? doesn't make sense to add a directory to a file path?

Reply:
Actually, `baseFile` is not a file path, but a file directory. `baseFile` is used in quite a few places, so I leave it to @kerimoyle to change it, if really needed.